### DISCUSSION TITLE
package age policy

### DISCUSSION THREAD

**Discussion by tennox** [renovatebot/renovate#12773] [Suggest an Idea]
### What would you like Renovate to be able to do?

In thinking about recent npm package malware I thought it would be great to be able to only install upgrades that are older than X (e.g. 1 month) - and have no reported security issues.

### If you have any ideas on how this should be implemented, please tell us here.

The only way I found will reading through existing issues and docs is maybe a custom version API that checks npm but only reports updates that meet the requirements.

But that would probably need custom implementation for each registry (e.g. docker, composer, ...)

So a package policy that checks minAge , when available would be way better.
Would be ready to implement when I get a rough sketch.

Or even custom filter plug-in / API that send package info and allows any custom check. But that's probably overkill.

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by viceice** [comment]
https://docs.renovatebot.com/configuration-options/#stabilitydays ?

**Comment by HonkingGoose** [comment]
We already have this feature, it's called `stabilityDays` [^docs]. 😉 

You can set `stabilityDays` to `31` so that you'll only ever get updates for packages that are a month old:

```json
{
  "packageRules": [
    {
      "stabilityDays": 31
    }
  ]
}
```

[^docs]: https://docs.renovatebot.com/configuration-options/#stabilitydays

**Comment by HonkingGoose** [comment]
If you're going to follow the route of "delay all updates for 31 days" I recommend you check that you have:

- [ ] Turned on the Dependency Graph (if using GitHub)
- [ ] Enabled Dependabot alerts (if using GitHub)

Or you may miss critical security updates.

https://docs.renovatebot.com/configuration-options/#vulnerabilityalerts

**Reply by tennox** [reply]
Thanks for the great advice!

**Reply by HonkingGoose** [reply]
You're welcome! Glad I could help a bit! 😄

**Comment by HonkingGoose** [comment]
> only install upgrades that are older than X (e.g. 1 month) - and have no reported security issues.

@rarkins Not sure if we want to keep this issue for the `and have no reported security issues` part?

I don't think Renovate currently checks things like GitHub Vulnerability Warnings or `npm audit` output when it creates a PR?

**Reply by rarkins** [reply]
I've converted this to a discussion as I don't see any functionality I'd like to add. The "no reported security issues" idea would not have any practical effect because malicious/malware packages such as the recent npm examples get yanked from the registry long before they show up in a vulnerability feed.

**Reply by tennox** [reply]
Fair enough :)

Is it worth to open a new issue for checking vulnerabilities when opening a PR, as @honkinggoose mentioned?

And could there be a possibility to get vulnerability alerts when not using GitHub?

**Reply by rarkins** [reply]
> Is it worth to open a new issue for checking vulnerabilities when opening a PR, as @HonkingGoose mentioned?

No, we don't plan to do that. It's of negligible benefit in practice and will add complexity

> And could there be a possibility to get vulnerability alerts when not using GitHub?

There's already an issue open for that

**Reply by HonkingGoose** [reply]
> There's already an issue open for that

You can subscribe to issue https://github.com/renovatebot/renovate/issues/6562 to be notified of any progress.

**Comment by Leksat** [comment]
Is there a way to configure Renovate like following?
- once a week create a single PR for all updates
- exclude major updates which are younger than X days (maybe just mention them in the PR)

What I try to achieve:
- Reduce the amount of PRs created by Renovate
- Reduce the CI load

As I far as I can get from the docs, the best I can do at the moment is:
```json5
  packageRules: [
    {
      matchUpdateTypes: ["minor", "patch", "digest"],
      groupName: "non-major",
    },
    {
      matchUpdateTypes: ["major"],
      stabilityDays: 14,
      prCreation: "not-pending",
    }
  ],
  dependencyDashboard: true,
```

But with that, Renovate still creates separate PRs for each major update, which increases CI load.

I saw that @ibratoev and @rarkins already discussed a possibility of creating an option to achieve the behavior I want in https://github.com/renovatebot/config-help/issues/572#issuecomment-604343934. But I could not find a follow-up for that discussion.

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#separatemajorminor

**Reply by Leksat** [reply]
@viceice but this does not help to get what I described. Maybe my explanation was poor, I'll try again 😅 

I'd like Renovate to create a _single_ PR for _all_ updates. But the PR should exclude major updates younger than X days.

This is not possible at the moment, right?

**Reply by viceice** [reply]
You need to disable the mentioned option, as it's enabled by default. That's why you get separate PR for major updates

**Reply by rarkins** [reply]
This advice is re: getting a single PR for all updates. Do that first, and then we can work on the update delay

**Reply by Leksat** [reply]
I already had a single PR for all updates. It worked quite well.

But when I added `stabilityDays`, the workflow broke. Because the PR included all updates, including the young major ones. And then Renovate was waiting for _all_ included major updates to pass the stabilityDays delay. So it was preventing the whole PR to be merged.

That's why I went with separating PRs to prevent unwanted delays:
1. have a separate PR for each major update
1. have a single PR for the rest of updates

The first point is required because otherwise Renovate can wait infinite amount of time. For example, if we still have all major updates in a single PR:
- major update for package A appeared - Renovate will wait 14 days before merging it
- after 7 days the base branch changed and Renovate retries the PR
- major update for package B appears in the PR - Renovate will wait 14 days again preventing package A to be updates

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#internalchecksfilter

**Reply by Leksat** [reply]
@viceice I probably did something wrong. Could you please take a look?

What I tried:
```json5
  packageRules: [
    {
      matchUpdateTypes: ["minor", "patch", "digest"],
      groupName: "all-non-major",
    },
    {
      matchUpdateTypes: ["major"],
      groupName: "all-major",
      stabilityDays: 5,
      internalChecksFilter: "strict",
    },
  ],
```

Renovate created a combined PR, but it still included major updates younger than 5 days (and added `renovate/stability-days` check).

(I read `internalChecksFilter` docs several times, but I still can't get how to use it right 😅)

**Reply by rarkins** [reply]
I think we'll need you to create a minimal reproduction repository

**Reply by Leksat** [reply]
Sorry for delay.

Here is the repo: https://github.com/Leksat/renovate-12773

Updates are grouped by major and non-major.

Currently the [all-major PR](https://github.com/Leksat/renovate-12773/pull/4) includes `gatsby`, `cypress` and `lint-staged` packages. The `renovate/stability-days` check is pending because `cypress` and `lint-staged` are younger than 30 days (`stabilityDays` value).

But instead I'd like the PR to only include passed packages. Which is `gatsby` in my case.

Is this feasible?

**Reply by rarkins** [reply]
Hi, I found that it works if you move the `internalChecksFilter` up to the top level: https://github.com/renovate-reproductions/renovate-12773/commit/425be14df94ade6dc3607bcbd1106d3bea7d23b5

The problem is that at the time the internalChecksFilter logic is executed, Renovate doesn't know yet the release type.

In addition I think you'd also be interested in #12914 so that you an filter based on the age of the major release stream and not just the age of the latest release in that stream (which will rarely be less than 30 days old, unless they release infrequently).

**Reply by Leksat** [reply]
Ooooooohhhhh... It works like a charm! Exactly what I wanted!

Guys, the level of support you provide is just FANTASTIC ❤️ ❤️ ❤️ Thank you!

One last question. Now `cypress v9` and `lint-staged v12` are not included in the PR, but they are also not mentioned anywhere. Can Renovate add a note somewhere that they were skipped due to internal checks?

**Reply by rarkins** [reply]
Unfortunately there's no way to do that today which I can think of

**Reply by Leksat** [reply]
Not a big deal 🙂 But would be a nice feature. Thanks!
